### PR TITLE
Something else

### PR DESCRIPTION
**Pull request by olekongithub**
Grab bag pull request, going to be working on discrete features from now on.
Some features:
-logging out after tests
-fixes in response to changes made by ofer
-stricter xpath selection
-general organization and trimming of code

### CODE REVIEW

File: tests/administration/realms.test.js [hunk 1]
```diff
@@ -1,7 +1,54 @@
...
 const common = require('../common'),
       expect = require('chai').expect;
 
 describe('Realms', () => {
   before(() => common.login(browser, 'admin', 'admin', 'local'));
+  after(() => common.logout());
   it('should navigate to Realms', () => common.clickSidebarTab(browser, 'Realms'));
+
+  it('should create a valid realm', () => {
+  	return common.waitForOnClick('//*[@id="add_realm"]')
```
Comment (mbroadst) [inline tests/administration/realms.test.js]: please keep to the existing style: 2 spaces for tabs

File: tests/administration/realms.test.js [hunk 2]
```diff
@@ -1,7 +1,54 @@
...
+  		.then(() => browser.click('//*[@id="add_realm"]'))
+  		.then(() => browser.waitUntil( browser.getAttribute('//*[@id="realm"]','onchange')
+  			.then((onc) => onc !== null )
+  		))
+  		.then(() => browser.setValue('//*[@id="realm"]', 'test'))
+  		.then(() => browser.waitUntil( browser.getAttribute('//*[@id="fqdn"]','onchange')
```
Comment (mbroadst) [inline tests/administration/realms.test.js]: no need to nest the `then` here, this should be:

```
.then(() => browser.waitUntil(browser.getAttribute('//*[@id="fqdn"]','onchange'))
.then(onc => onc !== null)
.then(...)
```

Comment (mbroadst) [inline tests/administration/realms.test.js]: oh I see sorry this is for the `waitUntil`, maybe just clean the style a little bit:

```
  .then(() => browser.waitUntil(browser.getAttribute('//*[@id="fqdn"]','onchange')
    .then((onc) => onc !== null)));
```

?

File: tests/administration/realms.test.js [hunk 3]
```diff
@@ -1,7 +1,54 @@
...
+  after(() => common.logout());
   it('should navigate to Realms', () => common.clickSidebarTab(browser, 'Realms'));
+
+  it('should create a valid realm', () => {
+  	return common.waitForOnClick('//*[@id="add_realm"]')
+  		.then(() => browser.click('//*[@id="add_realm"]'))
+  		.then(() => browser.waitUntil( browser.getAttribute('//*[@id="realm"]','onchange')
+  			.then((onc) => onc !== null )
+  		))
+  		.then(() => browser.setValue('//*[@id="realm"]', 'test'))
```
Comment (oferb1) [inline tests/administration/realms.test.js]: If the intent is to set a valid realm, then this should be HIVEIO - as we changed the page to indicate that the value needs to be the NetBIOS name of the domain.

Granted 'test' will pass the validation, but had we had validation of the NetBIOS name, it should have failed.

File: tests/administration/realms.test.js [hunk 4]
```diff
@@ -1,7 +1,54 @@
...
+  		.then(() => browser.waitUntil( browser.getAttribute('//*[@id="fqdn"]','onchange')
+  			.then((onc) => onc !== null )
+  		))
+  		.then(() => browser.setValue('//*[@id="fqdn"]','hiveio.local'))
+  		.then(() => browser.waitForExist('//*[@id="add_realm_form"]/button[1]'))
+  		.then(() => browser.click('//*[@id="add_realm_form"]/button[1]'))
+  		.then(() => browser.waitForExist('//*[@id="realm-message"]'))
+  		.then(() => browser.click('//*[@id="add_realm_form"]/button[1]'))
+  		.then(() => browser.waitForExist('//*[@id="page-wrapper"]//td[1 and text()="TEST"]'));
```
Comment (oferb1) [inline tests/administration/realms.test.js]: Likewise to the comment above this should be updated to HIVEIO

File: tests/administration/realms.test.js [hunk 5]
```diff
@@ -1,7 +1,54 @@
...
+
+  it('should return to page once add realm is cancelled', () => {
+  	return common.waitForOnClick('//*[@id="add_realm"]')
+  		.then(() => browser.click('//*[@id="add_realm"]'))
+  		.then(() => common.waitForOnClick('//*[@id="add_realm_form"]/button[2 and text()="Cancel"]'))
+  		.then(() => browser.click('//*[@id="add_realm_form"]/button[2 and text()="Cancel"]'))
+  		.then(() => browser.waitForExist('//*[@id="realm_form" and @class="hidden"]'))
+  });
+
+  it('should not allow an empty realm name', () => {
```
Comment (oferb1) [inline tests/administration/realms.test.js]: field changed to NetBIOS, test name should reflect

File: tests/administration/realms.test.js [hunk 6]
```diff
@@ -1,7 +1,54 @@
...
+  		))
+  		.then(() => browser.setValue('//*[@id="fqdn"]','hiveio.local'))
+  		.then(() => browser.waitForExist('//*[@id="add_realm_form"]/button[1]'))
+  		.then(() => browser.click('//*[@id="add_realm_form"]/button[1]'))
+  		.then(() => browser.waitForExist('//*[@id="realm-message"]'))
+  		.then(() => browser.click('//*[@id="add_realm_form"]/button[1]'))
+  		.then(() => browser.waitForExist('//*[@id="page-wrapper"]//td[1 and not(string())]', 1000, true));
+  });
+
 });
```
Comment (oferb1) [inline tests/administration/realms.test.js:94]: Missing a few tests:
- Test for both NetBIOS and FQDN being empty
- Test for validity of the FQDN, including:
  - Empty value
  - Format error (illegal FQDN)
  - A bogus FQDN
  - A value that isn't a domain controller

File: tests/administration/users.test.js [hunk 7]
```diff
@@ -3,54 +3,64 @@ const common = require('../common'),
...
+      .then(() => browser.setValue('//td[text()="admin" and position()=1]/..//form//input[@id="password"]', 'admin'))
+      .then(() => browser.waitUntil(
+        browser.getAttribute('//td[text()="admin" and position()=1]/..//form//button[@type="submit"]', 'type')
+          .then((cl) => { return cl !== null; })
+      ,10000))
+  	  .then(() => browser.click('//td[text()="admin" and position()=1]/..//form//button[@type="submit"]'))
       .then(() => browser.waitForExist('//h1[@class="page-header"]'))
       .then(() => browser.getText('//h1[@class="page-header"]'))
   	  .then((text) => expect(text).to.equal('System Users'));
   });
```
Comment (oferb1) [inline tests/administration/users.test.js]: Change password test should probably be more complete, and include, changing it to something that isn't the same default password -> logout -> validate the password was set correctly, and then change it back to the default.

This can also be done by opening a new window/tab, and validate successful login

File: tests/administration/users.test.js [hunk 8]
```diff
@@ -3,54 +3,64 @@ const common = require('../common'),
...
-      .then(() => browser.click('//td[contains(text(),"test") and position()=1]/..//button[contains(@id,"rm")]'))
-      .then(() => browser.waitForExist('//td[contains(text(),"test") and position()=1]/..//button[contains(@id,"rm")]', 500, true))
+      .then(() => browser.waitForExist('//td[text()="test" and position()=1]/..//button[contains(@id,"rm")]'))
+      .then(() => browser.waitUntil(
+        browser.getAttribute('//td[text()="test" and position()=1]/..//button[contains(@id,"rm")]', 'onclick')
+          .then((cl) => { return cl !== null; })
+        ,10000))
+      .then(() => browser.click('//td[text()="test" and position()=1]/..//button[contains(@id,"rm")]'))
+      .then(() => browser.waitForExist('//td[text()="test" and position()=1]/..//button[contains(@id,"rm")]', 5000, true))
   });
```
Comment (oferb1) [inline tests/administration/users.test.js]: As above, if we do not verify the newly created user, we just know that the form doesn't spit an error

File: tests/administration/users/new_user_validation.test.js [hunk 9]
```diff
@@ -3,37 +3,19 @@ const common = require('../../common'),
       expect = require('chai').expect;
```
Comment (oferb1) [inline tests/administration/users/new_user_validation.test.js]: Not sure why this was separated to a different file...

File: tests/administration/users.test.js [hunk 10]
```diff
@@ -60,18 +70,12 @@ describe('Users', () => {
...
   it('should not allow deletion of current admin user', () => {
-    return browser.waitForExist(`//td[contains(text(),"admin") and position()=1]/..//button[contains(@id,"rm")]`, 500, true)
+    return browser.waitForExist(`//td[text()="admin" and position()=1]/..//button[contains(@id,"rm")]`, 5000, true)
   });
-
-  // it.only('should log out', () => {
-  //   return common.logout()
-  // })
 });
-
```
Comment (oferb1) [inline tests/administration/users.test.js]: Missing:
- Should not allow changing the password to an empty password (will fail in b22)
- Realm user testing

File: tests/administration/users/new_user_validation.test.js [hunk 11]
```diff
@@ -3,37 +3,19 @@ const common = require('../../common'),
...
 
 describe('New User Validations', () => {
-  before(() => { 
-	common.login(browser, 'admin', 'admin', 'local' )
-	.then(() => common.clickSidebarTab(browser, 'Users', 'System Users'));
-  });
+  beforeEach(() => common.login(browser, 'admin', 'admin', 'local' )
+  	.then(() => common.clickSidebarTab(browser, 'Users', 'System Users')));
```
Comment (oferb1) [inline tests/administration/users/new_user_validation.test.js]: Do we really need to log in and log out after every test? the tests that look for validation errors on a field within the form  don't really require that.

i guess this goes to for the previous and next test files... perhaps we should have a way to handle tests that require this and tests that don't (if possible)

File: tests/common.js [hunk 12]
```diff
@@ -1,61 +1,100 @@
...
       .then(() => browser.click('//button[@type="submit"]'))
-
-      // wait for the page to show up, default page is the "Overview"
-      .then(() => browser.waitForExist('h1.page-header'))
-      .then(() => browser.getText('h1.page-header'))
-      .then(header => expect(header).to.equal('Overview'));
+      .then(() => browser.waitForExist('//h1[@class="page-header"]'))
+      .then(() => browser.waitUntil(() => 
+        browser.getText('//h1[@class="page-header"]')
+        .then((text) => { return text === 'Overview'; })
```
Comment (oferb1) [inline tests/common.js]: Not sure I like the success criteria - and if we do use it, I suggest we set it as a variable, that defaults to 'Overview' - if we want to use the same common for testing of other sites, like the broker, the landing page will be different...

alternatively, look for different success criteria (session cookie? JWT down the road, etc.)

File: tests/dashboard/guest-management.test.js [hunk 13]
```diff
@@ -3,6 +3,7 @@ const common = require('../common'),
 
 describe('Guest Management', () => {
   before(() => common.login(browser, 'admin', 'admin', 'local'));
+  after(() => common.logout());
   it('should navigate to Guest Management', () =>
     common.clickSidebarTab(browser, 'Guest Management'));
 });
```
Comment (oferb1) [inline tests/dashboard/guest-management.test.js:9]: This one will have to be revisited once you have the full lifecycle testing - as successfully testing this page requires either a laundry list of prerequisites (template, pool, and potentially realm).

In addition, some of the test cases might require injection of canned records to the DB. for example, the options that are enabled in a guest's dropdown is related to it's "status", for example, soft power command (reboot/shutdown) should only be enabled for 'Ready', 'Activated' and, 'Deactivated' states, undefine only enabled for 'Shut Off' state, etc.

File: tests/dashboard/network-graphs.test.js [hunk 14]
```diff
@@ -26,5 +27,4 @@ describe('Network Graphs', () => {
       .then(title => expect(title).to.equal(test.expected));
     });
   });
-
```
Comment (oferb1) [inline tests/dashboard/network-graphs.test.js:29]: Should include:
- Are all network interfaces present in the drop down?
- Do they all have a graph
- Change of interface in dropdown generates a new graph
- Validate that data streams in (e.g. should wait a for the graph to update)
- Validate empty graphs - need to think how this is possible

File: tests/dashboard/overview.test.js [hunk 15]
```diff
@@ -3,6 +3,7 @@ const common = require('../common'),
 
 describe('Overview', () => {
   before(() => common.login(browser, 'admin', 'admin', 'local'));
+  after(() => common.logout());
   it('should navigate to Overview', () =>
     common.clickSidebarTab(browser, 'Overview'));
 });
```
Comment (oferb1) [inline tests/dashboard/overview.test.js:15]: Should include:
- Are all gauges displayed
- Do they update?
- Validate data in the overview table (as much as possible, or that at least it is displayed and passes a validation)

File: tests/dashboard/system-graphs.test.js [hunk 16]
```diff
@@ -3,8 +3,9 @@ const common = require('../common'),
```
Comment (oferb1) [inline tests/dashboard/system-graphs.test.js:7]: Missing a test if the data updates, and (again if possible) verification that the graphs are not empty

File: tests/common.js [hunk 17]
```diff
@@ -1,61 +1,100 @@
+'use strict';
 const chai = require('chai'), expect = chai.expect;
 chai.use(require('chai-string'));
 
 module.exports = {
+
   login: function(browser, username, password, domain) {
-    return browser.url('/').getTitle()
```
Comment (oferb1) [inline tests/common.js:6]: Should we not test that we are not already logged in?

File: tests/login.test.js [hunk 18]
```diff
@@ -19,5 +19,4 @@ describe('log in to the admin page', () => {
       .then(() => browser.getText('h1.page-header'))
       .then(header => expect(header).to.equal('Overview'));
   });
-
 });
```
Comment (oferb1) [inline tests/login.test.js]: - Any reason this is not using the common login function?
- Potentially we should probably look into LDAP authentication as well, so not just local
- Test readonly, etc.
- Logout at the end, etc.

File: tests/publishing/convert-image.test.js [hunk 19]
```diff
@@ -5,19 +5,34 @@ describe('Convert Image', () => {
...
+  });
+
+  it('should clear old output error messages', () => {
+    return browser.setValue('//input[@id="output"]', 'invalid input')
+      .then(() => browser.click('//button[@id="btn_convert"]'))
+      .then(() => browser.setValue('//input[@id="output"]', ''))
+      .then(() => browser.click('//button[@id="btn_convert"]'))
+      .then(() => browser.waitForExist('//input[@id="output"]/..//i[position()=2]',500,true))
+  });      
 });
```
Comment (oferb1) [inline tests/publishing/convert-image.test.js:45]: If the conversion is done on a large file, the current implementation, will eventually stop the spinner, and alert the user that the conversion is still running (this is due to the shitty way it is implemented, that causes the underlying Nginx session to timeout) - not sure we want to test that, just bringing it up...
- noticed that not conversion test actually takes place

File: tests/publishing/guest-pools.test.js [hunk 20]
```diff
@@ -1,8 +1,95 @@
...
+  	 return browser.waitForExist('//*[@id="add_pool"]')
+  		.then(() => browser.isExisting('//*[@id="add_pool" and contains(@class,"disabled")]'))
+  		.then((ex) => {if(!ex){ 
+  			return common.waitAndClick('//*[@id="add_pool"]')
+  				.then(() => common.waitAndClick('//*[@id="add_pool_form"]/div[8]/button[2 and text()="Cancel"]'))
+					.then(() => browser.waitForExist('//*[@id="add_pool" and not(contains(@class,"hidden"))]')); 		
+			}});
+  });
+
+  it('should allow creating a new pool when a template exists', () => {
```
Comment (oferb1) [inline tests/publishing/guest-pools.test.js]: When is the template created?

File: tests/publishing/templates.test.js [hunk 21]
```diff
@@ -4,28 +4,34 @@ const common = require('../common'),
...
+  });
 
+  it('should allow user to cancel adding a template', () => {
+    return common.waitForOnClick('//*[@id="add_tmpl_form"]/div[3]/button[2 and text()="Cancel"]')
+      .then(() => browser.click('//*[@id="add_tmpl_form"]/div[3]/button[2 and text()="Cancel"]'))
+      .then(() => browser.waitForExist('//*[@id="tmpl_form" and @class="hidden"]'));
+  });
 
 });
-
```
Comment (oferb1) [inline tests/publishing/templates.test.js:29]: Should probably go under the tests/publishing/templates/ folder, and be called basic.test.js

Should test successful scenarios as well, and place them in the common.js as we will have tests down the road that require a valid template.

Also seems to be missing all the workload around the new template, authoring etc. 

Lastly - we are missing functionality that will manage the datastore, as we will want tests that create a new template, and should then be able to delete it, test the functionality of authoring, including the .authoring file, and be able to delete said file is it is stuck.

File: tests/publishing/templates/validations.test.js [hunk 22]
```diff
@@ -7,44 +7,77 @@ function loginAndClickTemplates() {
...
-      .then(() => browser.click(`${form}//input[@name="name"]`))
+  it('should refuse an invalid template path in new template', () => {
+    return common.waitForOnClick('//button[@id="new_tmpl"]')
+      .then(() => browser.click('//button[@id="new_tmpl"]'))
+      .then(() => browser.setValue('//form[@id="newTmplForm"]//input[@id="path"]', 'invalid input'))
+      .then(() => browser.click('//form[@id="newTmplForm"]//input[@name="name"]'))
       .then(() => browser.getText('//span[@id="path-message"]'))
-      .then((text) => expect(text).to.equal("Path verification failed - can't mount destination."))
+      .then((text) => expect(text).to.equal("Path verification failed - can't mount destination."));
   });
```
Comment (oferb1) [inline tests/publishing/templates/validations.test.js:29]: If this test fails, need to clean it up, as it may cause issues for tests down the road - this is true to many other tests for templates, pools, users, pools, etc.

File: tests/publishing/templates/validations.test.js [hunk 23]
```diff
@@ -7,44 +7,77 @@ function loginAndClickTemplates() {
...
 
-  // it('puts', () => {
-  //   expect(true).to.be.true;
-  // });
+  it('should refuse a new template form without a path', () => {
```
Comment (oferb1) [inline tests/publishing/templates/validations.test.js]: Harping on the same logout after the above (that we expect to fail and make no change in the system), to then login and start this test that we also expect to make no changes in the system

File: tests/administration/realms.test.js (hunk 1 above)
Comment (olekongithub) [inline tests/administration/realms.test.js]: done

File: tests/administration/realms.test.js (hunk 3 above)
Comment (olekongithub) [inline tests/administration/realms.test.js]: done

File: tests/publishing/templates/validations.test.js [hunk 24]
```diff
@@ -7,44 +7,77 @@ function loginAndClickTemplates() {
...
 
+  it('should refuse an add template form with a name and an invalid path', () => {
+    return common.waitAndClick('//*[@id="add_tmpl"]')
+      .then(() => browser.setValue('//*[@id="name"]', 'All Hail Slithis!'))
+      .then(() => browser.setValue('//*[@id="path"]', 'All Hail Slithis!'))
+      .then(() => common.waitAndClick('//*[@id="subBtn"]'))
+      .then(() => browser.waitForExist('//*[@id="path-message"]', 10000));
+  });
 
+});
```
Comment (oferb1) [inline tests/publishing/templates/validations.test.js:86]: This is an example of an entire test file that does not really require a logout/login between the tests, as these are all validation tests that should not have an effect on the system or that require a "clean slate"

File: tests/administration/realms.test.js (hunk 2 above)
Comment (olekongithub) [inline tests/administration/realms.test.js]: done

File: tests/administration/realms.test.js (hunk 4 above)
Comment (olekongithub) [inline tests/administration/realms.test.js]: done

File: tests/administration/realms.test.js (hunk 5 above)
Comment (olekongithub) [inline tests/administration/realms.test.js]: done

File: tests/administration/realms.test.js (hunk 6 above)
Comment (olekongithub) [inline tests/administration/realms.test.js:94]: could i get examples of the last three? illegal, bogus and not a domain controller

File: tests/administration/realms.test.js [hunk 25]
```diff
@@ -1,7 +1,54 @@
...
+  		))
+  		.then(() => browser.setValue('//*[@id="fqdn"]','hiveio.local'))
+  		.then(() => browser.waitForExist('//*[@id="add_realm_form"]/button[1]'))
+  		.then(() => browser.click('//*[@id="add_realm_form"]/button[1]'))
+  		.then(() => browser.waitForExist('//*[@id="realm-message"]'))
+  		.then(() => browser.click('//*[@id="add_realm_form"]/button[1]'))
+  		.then(() => browser.waitForExist('//*[@id="page-wrapper"]//td[1 and text()="TEST"]'));
+  });
+
+  it('should return to page once add realm is cancelled', () => {
```
Comment (oferb1) [inline tests/administration/realms.test.js]: With the current log in/log off - this test will never really be "after realm is cancelled" as the system will log off, and the log in again...

File: tests/administration/realms.test.js (hunk 6 above)
Comment (olekongithub) [inline tests/administration/realms.test.js:94]: added a to do list to realms to make these tests

File: tests/administration/users.test.js (hunk 7 above)
Comment (olekongithub) [inline tests/administration/users.test.js]: password change tests now log in and back out

File: tests/administration/users.test.js (hunk 8 above)
Comment (olekongithub) [inline tests/administration/users.test.js]: done

File: tests/administration/users.test.js (hunk 10 above)
Comment (olekongithub) [inline tests/administration/users.test.js]: added change to empty password test, realm test will wait until integrated scenario tests are developed

File: tests/administration/users/new_user_validation.test.js (hunk 11 above)
Comment (olekongithub) [inline tests/administration/users/new_user_validation.test.js]: removed this

File: tests/common.js (hunk 17 above)
Comment (olekongithub) [inline tests/common.js:6]: now testing that we are not logged in

File: tests/common.js (hunk 12 above)
Comment (olekongithub) [inline tests/common.js]: user doesn't have access to his own session cookie, leaving it as is for now.

File: tests/dashboard/guest-management.test.js (hunk 13 above)
Comment (olekongithub) [inline tests/dashboard/guest-management.test.js:9]: just the setup for testing guest pools will be more complicated than any of my current tests

File: tests/dashboard/network-graphs.test.js (hunk 14 above)
Comment (olekongithub) [inline tests/dashboard/network-graphs.test.js:29]: added to to do list in network graphs

File: tests/dashboard/overview.test.js (hunk 15 above)
Comment (olekongithub) [inline tests/dashboard/overview.test.js:15]: added to to do list in overview.

File: tests/login.test.js (hunk 18 above)
Comment (olekongithub) [inline tests/login.test.js]: removed this test as we are already using it a lot throughout our tests 

readonly is tested in users

logout is used throughout

will revisit once I get it set up for realm login

File: tests/publishing/convert-image.test.js (hunk 19 above)
Comment (olekongithub) [inline tests/publishing/convert-image.test.js:45]: no conversion yet, will implement when we have a stable of convertible files on deathstar
added spinner bug to to do list in test

File: tests/publishing/guest-pools.test.js (hunk 20 above)
Comment (olekongithub) [inline tests/publishing/guest-pools.test.js]: it isn't. Was just putting this in by hand. I will be moving all of this to a scenario.

File: tests/publishing/templates.test.js (hunk 21 above)
Comment (olekongithub) [inline tests/publishing/templates.test.js:29]: moved it and changed the name.

creating templates, authoring etc will come in the next branch

datastore management is even further down the line but we will get there.

File: tests/publishing/templates/validations.test.js (hunk 22 above)
Comment (olekongithub) [inline tests/publishing/templates/validations.test.js:29]: I am adding clean ups to all tests that can change the state of hio

File: tests/publishing/templates/validations.test.js (hunk 24 above)
Comment (olekongithub) [inline tests/publishing/templates/validations.test.js:86]: it only had a before not beforeEach in this one so it was not logging in and out between each test.

File: tests/dashboard/system-graphs.test.js (hunk 16 above)
Comment (olekongithub) [inline tests/dashboard/system-graphs.test.js:7]: added to do list
